presto-main/src/main/java/com/facebook/presto/operator/PageBufferClient.java

```diff
@@ -197,23 +196,17 @@ public synchronized boolean isRunning()
     public void close()
     {
         boolean shouldSendDelete;
-        Future<?> future;
         synchronized (this) {
             shouldSendDelete = !closed;
 
             closed = true;
 
-            future = this.future;
-
+            // do not cancel the request to avoid closing a pooled connection
```

Comment:
Is this the correct semantic when the RpcShuffleClient is Thrift based?  More generally, should this change be implemented in the HttpClient (the future it returns won't support cancellation?)

Reply:
That's the way it is implemented in Drift. Basically the future it returns is not cancelled (cancellation of the future does not cancel the request). 
It has it's own pros and cons.
Basically the pros are that the connection is never closed, the cons is that if the request is large - it is never cancelled, and in theory the request / response can be in megabytes.

Reply:
I'm thinking about a compromise. The `Future#cancel()` method takes a boolean parameter, the `mayInterruptIfRunning`. Currently this flag is unused, and no matter what's the value the request is always cancelled. What do you think about changing this behavior to "cancel only if the mayInterruptIfRunning flag is set"?

e.g.: https://github.com/prestodb/airlift/blob/master/http-client/src/main/java/com/facebook/airlift/http/client/jetty/JettyResponseFuture.java#L59

Reply:
When it's becoming even more interesting - is when the HTTP2 comes into play. With HTTP/2 it is possible to terminate the request without closing the connection. Thus the decision whether to terminate or not to terminate depends on the underlying protocol as well. 

But anyhow, I don't think there's a real downside for Presto specifically in not terminating requests. The requests are small, and the termination is inherently racy. Thus I would argue it is fine to not to terminate the requests for our usecase.

Reply:
@tdcmeehan @wenleix 

How about next proposal:

1. I change the implementation of the `Future#cancel()` in Airlift so when `mayInterruptIfRunning` set to false - the request won't be terminated.

Reply:
I think it's reasonable.

Reply:
https://github.com/prestodb/airlift/pull/16